src/main/scala/com/indix/mesos/GOCDPoller.scala

```diff
@@ -28,19 +52,48 @@ case class GOCDPoller(conf: FrameworkConfig) {
   }
 
   def goIdleAgentsCount() = {
+    println("Polling Go server for idle agents")
+    try {
+      val responseJson = jsonRequest(buildUrl() + "/go/api/agents")
+      (responseJson \ "_embedded" \ "agents").toOption.map(jsValue => {
+        jsValue.as[JsArray].value.map(agent => (agent \ "status").get.as[String]).count(_.equalsIgnoreCase("idle"))
+      }).getOrElse(0)
+    } catch {
+      case e: SocketTimeoutException => {
+        println("GOCD Server timed out!!")
+        1
+      }
+    }
+  }
+
+  // Return true when the demand remains, same or increased during the last five attempts.
+  private[mesos] def isDemandPositive(): Boolean = {
+    val demandTrend = responseHistory.foldRight(0)((h1, h2) => {
+     if(h2 >= h1) {
+       1
+     } else {
+       -1
+     }
+    })
+    demandTrend == 1
   }
 
   def pollAndAddTask() = {
     val scheduled : Int = goTaskQueueSize()
     println(s"Go server has ${scheduled.toString} pending jobs to be scheduled")
-    if(scheduled > 0)
+    if(scheduled > 0) {
       responseHistory += scheduled
-    
+    }
     if(responseHistory.size > 5) {
-      println(s"More than 5 jobs pending in the GOCD. queuing a new agent launch now.")
-      TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
-      responseHistory.clear()
+      println(s"More than 5 jobs pending in the GOCD. checking if any agents are idle.")
+      if(goIdleAgentsCount() == 0 && isDemandPositive()) {
+        println("No idle agents found. Launching a new agent launch now.")
+        TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
+        responseHistory.clear()
+      } else {
+        println("Idle agents found. Not launching any new Go Agent now.")
+        responseHistory.drop(0)
```

Comment:
So, this is what I had in my mind:
- Poll for job queue status
- If there's an upscale action needed, determine how many agents need to be
brought up.
- Issue scale event for the # of agents
- Mark the Scalar state to UPSCALING
- Meanwhile, polling continues, but, use it only to update history, not to
trigger events.
- Mark Scalar state to IDLE once upscaling completes
- Everything continues to chug along



On Sat, Feb 20, 2016 at 8:56 PM, Sriram Ramachandrasekaran <sriram@indix.com
> wrote:

> Generally, that's handled via "UPSCALING" state in your GoCD mesos
> framework.
>
> On Sat, Feb 20, 2016 at 8:54 PM, Azhaguselvan SP <notifications@github.com
> > wrote:
>
>> In src/main/scala/com/indix/mesos/GOCDPoller.scala
>> <https://github.com/ind9/gocd-mesos/pull/3#discussion_r53551958>:
>>
>> >      if(responseHistory.size > 5) {
>> > -      println(s"More than 5 jobs pending in the GOCD. queuing a new agent launch now.")
>> > -      TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
>> > -      responseHistory.clear()
>> > +      println(s"More than 5 jobs pending in the GOCD. checking if any agents are idle.")
>> > +      if(goIdleAgentsCount() == 0 && isDemandPositive()) {
>> > +        println("No idle agents found. Launching a new agent launch now.")
>> > +        TaskQueue.enqueue(GoTask("", conf.goAgentDocker, ""))
>> > +        responseHistory.clear()
>> > +      } else {
>> > +        println("Idle agents found. Not launching any new Go Agent now.")
>> > +        responseHistory.drop(0)
>>
>> @sriram-r <https://github.com/sriram-r> Hmmm. Once a new agent is
>> launched, we don't expect the jobs to get scheduled immediately. There is
>> some lag between that. If we don't clean the history, isn't there is a
>> possibility that before this agent is launched and the job is scheduled,
>> another one might also get launched?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ind9/gocd-mesos/pull/3/files#r53551958>.
>>
>
>